### PR TITLE
Add support for binding to a topic pattern for streams binder

### LINKED ISSUE #820: Kafka Streams Binder fails when destination is a pattern
I tried to write a Kafka Streams function using Spring Cloud Stream and the Kafka Streams Binder. The destination of the input needs to be a pattern as the same processing shall happen for a bunch of different topics.

The application starts successfully but the log output contains errors while initializing the Binding:

```
2019-12-16 08:27:40.592  INFO 7801 --- [  restartedMain] o.a.kafka.common.utils.AppInfoParser     : Kafka version: 2.3.1
2019-12-16 08:27:40.592  INFO 7801 --- [  restartedMain] o.a.kafka.common.utils.AppInfoParser     : Kafka commitId: 18a913733fb71c01
2019-12-16 08:27:40.592  INFO 7801 --- [  restartedMain] o.a.kafka.common.utils.AppInfoParser     : Kafka startTimeMs: 1576481260592
2019-12-16 08:27:40.596  INFO 7801 --- [  restartedMain] org.apache.kafka.streams.KafkaStreams    : stream-client [data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30] State transition from CREATED to REBALANCING
2019-12-16 08:27:40.597  INFO 7801 --- [-StreamThread-1] o.a.k.s.p.internals.StreamThread         : stream-thread [data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1] Starting
2019-12-16 08:27:40.597  INFO 7801 --- [-StreamThread-1] o.a.k.s.p.internals.StreamThread         : stream-thread [data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1] State transition from CREATED to STARTING
2019-12-16 08:27:40.598  INFO 7801 --- [-StreamThread-1] o.a.k.clients.consumer.KafkaConsumer     : [Consumer clientId=data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1-consumer, groupId=data-transformer] Subscribed to pattern: 'etl.processing.*'
2019-12-16 08:27:40.617  INFO 7801 --- [-StreamThread-1] org.apache.kafka.clients.Metadata        : [Consumer clientId=data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1-consumer, groupId=data-transformer] Cluster ID: KpwWiStVRIuc5fbEd37Edw
2019-12-16 08:27:40.618  INFO 7801 --- [-StreamThread-1] o.a.k.c.c.internals.AbstractCoordinator  : [Consumer clientId=data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1-consumer, groupId=data-transformer] Discovered group coordinator stoetti-workstation:9092 (id: 2147483647 rack: null)
2019-12-16 08:27:40.619  INFO 7801 --- [-StreamThread-1] o.a.k.c.c.internals.ConsumerCoordinator  : [Consumer clientId=data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1-consumer, groupId=data-transformer] Revoking previously assigned partitions []
2019-12-16 08:27:40.619  INFO 7801 --- [-StreamThread-1] o.a.k.s.p.internals.StreamThread         : stream-thread [data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1] State transition from STARTING to PARTITIONS_REVOKED
2019-12-16 08:27:40.620  INFO 7801 --- [-StreamThread-1] o.a.k.clients.consumer.KafkaConsumer     : [Consumer clientId=data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1-restore-consumer, groupId=null] Unsubscribed all topics or patterns and assigned partitions
2019-12-16 08:27:40.620  INFO 7801 --- [-StreamThread-1] o.a.k.s.p.internals.StreamThread         : stream-thread [data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1] partition revocation took 0 ms.
	suspended active tasks: []
	suspended standby tasks: []
2019-12-16 08:27:40.620  INFO 7801 --- [-StreamThread-1] o.a.k.c.c.internals.AbstractCoordinator  : [Consumer clientId=data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1-consumer, groupId=data-transformer] (Re-)joining group
2019-12-16 08:27:40.629  INFO 7801 --- [-StreamThread-1] o.a.k.c.c.internals.AbstractCoordinator  : [Consumer clientId=data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1-consumer, groupId=data-transformer] (Re-)joining group
2019-12-16 08:27:40.633 ERROR 7801 --- [-StreamThread-1] o.a.k.s.p.i.StreamsPartitionAssignor     : stream-thread [data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1-consumer] Missing source topic etl.processing.* durign assignment. Returning error INCOMPLETE_SOURCE_TOPIC_METADATA.
2019-12-16 08:27:40.633 ERROR 7801 --- [-StreamThread-1] o.a.k.s.p.i.StreamsPartitionAssignor     : stream-thread [data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1-consumer] etl.processing.* is unknown yet during rebalance, please make sure they have been pre-created before starting the Streams application.
2019-12-16 08:27:40.634  WARN 7801 --- [-StreamThread-1] o.a.k.c.c.internals.ConsumerCoordinator  : [Consumer clientId=data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1-consumer, groupId=data-transformer] The following subscribed topics are not assigned to any members: [etl.processing.shell] 
2019-12-16 08:27:40.637  INFO 7801 --- [-StreamThread-1] o.a.k.c.c.internals.AbstractCoordinator  : [Consumer clientId=data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1-consumer, groupId=data-transformer] Successfully joined group with generation 9
2019-12-16 08:27:40.638  INFO 7801 --- [-StreamThread-1] o.a.k.c.c.internals.ConsumerCoordinator  : [Consumer clientId=data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1-consumer, groupId=data-transformer] Setting newly assigned partitions: 
2019-12-16 08:27:40.638 ERROR 7801 --- [-StreamThread-1] o.a.k.s.p.internals.StreamThread         : stream-thread [data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1] Received error code 1 - shutdown
2019-12-16 08:27:40.639  INFO 7801 --- [-StreamThread-1] o.a.k.s.p.internals.StreamThread         : stream-thread [data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1] Informed to shut down
2019-12-16 08:27:40.639  INFO 7801 --- [-StreamThread-1] o.a.k.s.p.internals.StreamThread         : stream-thread [data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1] State transition from PARTITIONS_REVOKED to PENDING_SHUTDOWN
2019-12-16 08:27:40.659  INFO 7801 --- [read-1-producer] org.apache.kafka.clients.Metadata        : [Producer clientId=data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1-producer] Cluster ID: KpwWiStVRIuc5fbEd37Edw
2019-12-16 08:27:40.660  INFO 7801 --- [  restartedMain] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat started on port(s): 8080 (http) with context path ''
2019-12-16 08:27:40.661  INFO 7801 --- [  restartedMain] c.u.o.f.s.DataTransformerApplication     : Started DataTransformerApplication in 5.145 seconds (JVM running for 5.964)
2019-12-16 08:27:40.706  INFO 7801 --- [-StreamThread-1] o.a.k.s.p.internals.StreamThread         : stream-thread [data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1] Shutting down
2019-12-16 08:27:40.707  INFO 7801 --- [-StreamThread-1] o.a.k.clients.consumer.KafkaConsumer     : [Consumer clientId=data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1-restore-consumer, groupId=null] Unsubscribed all topics or patterns and assigned partitions
2019-12-16 08:27:40.707  INFO 7801 --- [-StreamThread-1] o.a.k.clients.producer.KafkaProducer     : [Producer clientId=data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1-producer] Closing the Kafka producer with timeoutMillis = 9223372036854775807 ms.
2019-12-16 08:27:40.737  INFO 7801 --- [-StreamThread-1] o.a.k.s.p.internals.StreamThread         : stream-thread [data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1] State transition from PENDING_SHUTDOWN to DEAD
2019-12-16 08:27:40.737  INFO 7801 --- [-StreamThread-1] org.apache.kafka.streams.KafkaStreams    : stream-client [data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30] State transition from REBALANCING to ERROR
2019-12-16 08:27:40.737 ERROR 7801 --- [-StreamThread-1] org.apache.kafka.streams.KafkaStreams    : stream-client [data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30] All stream threads have died. The instance will be in error state and should be closed.
2019-12-16 08:27:40.737  INFO 7801 --- [-StreamThread-1] o.a.k.s.p.internals.StreamThread         : stream-thread [data-transformer-4d05daff-774a-46a5-b5c6-a3d9a8730a30-StreamThread-1] Shutdown complete
```

Are there any limitations using the combination of Kafka Streams Binder and destination patterns?

I am using Spring Boot 2.2.2 and Spring Cloud Hoxton.Release

### PR DESCRIPTION
**Pull request by stoetti**
Streams Binder did not support usage of topic patterns as described in #820 

Resolves  #820

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@stoetti Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-stream-binder-kafka&pullRequestId=821)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-cloud/spring-cloud-stream-binder-kafka&pullRequestId=821) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@stoetti Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-stream-binder-kafka&pullRequestId=821)!

**Comment by sobychacko** [comment]
@stoetti Thank you for the PR. This is merged upstream.

File: spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/AbstractKafkaStreamsBinderProcessor.java [hunk 1]
```diff
@@ -279,16 +284,24 @@ else if (deserializationExceptionHandler == DeserializationExceptionHandler.send
...
-		String[] bindingTargets = StringUtils.commaDelimitedListToStringArray(
-				this.bindingServiceProperties.getBindingDestination(inboundName));
-		final Consumed<?, ?> consumed = getConsumed(kafkaStreamsConsumerProperties, keySerde, valueSerde, autoOffsetReset);
-		KStream<?, ?> stream = streamsBuilder.stream(Arrays.asList(bindingTargets),
-				consumed);
+		KStream<?, ?> stream;
+		if (this.kafkaStreamsExtendedBindingProperties
+				.getExtendedConsumerProperties(inboundName).isDestinationIsPattern()) {
+			final Pattern pattern = Pattern.compile(this.bindingServiceProperties.getBindingDestination(inboundName));
+			stream = streamsBuilder.stream(pattern);
```
Comment (ples) [inline spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/AbstractKafkaStreamsBinderProcessor.java:296]: @stoetti Why serde's args are missing here? Is it conflicting with topic pattern approach?

Comment (stoetti) [reply spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/AbstractKafkaStreamsBinderProcessor.java:296]: There might be an issue with serde when registering for message from an unknown set of topics as new topics may be created that use a different schema or data type in the payload.

**Pull request closed without merging**